languages/cpp2/src/main/java/de/jplag/cpp2/CPPTokenListener.java

```diff
@@ -0,0 +1,359 @@
+package de.jplag.cpp2;
+
+import java.util.ArrayDeque;
+import java.util.Set;
+
+import org.antlr.v4.runtime.ParserRuleContext;
+import org.antlr.v4.runtime.Token;
+import org.antlr.v4.runtime.tree.ParseTree;
+
+import de.jplag.TokenType;
+import de.jplag.cpp2.grammar.CPP14Parser;
+import de.jplag.cpp2.grammar.CPP14ParserBaseListener;
+
+/**
+ * Extracts tokens from the ANTLR parse tree. Token extraction is built to be similar to the Java language module. In
+ * some cases, the grammar is ambiguous and requires surrounding context to extract the correct token. Those cases are
+ * covered by {@link #enterSimpleTypeSpecifier(CPP14Parser.SimpleTypeSpecifierContext)} and
+ * {@link #enterSimpleDeclaration(CPP14Parser.SimpleDeclarationContext)}.
+ */
+public class CPPTokenListener extends CPP14ParserBaseListener {
+
+    private final CPPParserAdapter parser;
+
+    public CPPTokenListener(CPPParserAdapter parser) {
+        this.parser = parser;
+    }
+
+    @Override
+    public void enterClassSpecifier(CPP14Parser.ClassSpecifierContext context) {
+        if (context.classHead().Union() != null) {
+            addEnter(CPPTokenType.UNION_BEGIN, context.getStart());
```

Comment:
You could use static imports here and at other occurrences for the token types; this would increase readability.